src/compiler/checker.ts

```diff
@@ -45073,6 +45087,44 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         return value;
     }
 
+    function evaluatesToNumericLiteral(expr: Expression): boolean {
+        switch (expr.kind) {
+            case SyntaxKind.PrefixUnaryExpression:
+                return evaluatesToNumericLiteral((expr as PrefixUnaryExpression).operand);
+            case SyntaxKind.BinaryExpression:
+                return evaluatesToNumericLiteral((expr as BinaryExpression).left) && evaluatesToNumericLiteral((expr as BinaryExpression).right);
+            case SyntaxKind.ParenthesizedExpression:
+                return evaluatesToNumericLiteral((expr as ParenthesizedExpression).expression);
+            case SyntaxKind.NumericLiteral:
+                return true;
+        }
+        return false;
+    }
+
+    function evaluatesToStringLiteral(expr: Expression): boolean {
+        switch (expr.kind) {
+            case SyntaxKind.BinaryExpression:
+                const left = (expr as BinaryExpression).left;
+                const right = (expr as BinaryExpression).right;
+                const leftIsNumeric = evaluatesToNumericLiteral(left);
+                const rightIsNumeric = evaluatesToNumericLiteral(right);
+                return (
+                    !(leftIsNumeric && rightIsNumeric) &&
+                    (evaluatesToStringLiteral(left) || leftIsNumeric) &&
+                    (evaluatesToStringLiteral(right) || rightIsNumeric) &&
+                    (expr as BinaryExpression).operatorToken.kind === SyntaxKind.PlusToken
+                );
+            case SyntaxKind.TemplateExpression:
+                return (expr as TemplateExpression).templateSpans.every(span => evaluatesToStringLiteral(span.expression));

```

Comment:
In this PR I wanted to implement https://github.com/microsoft/TypeScript/issues/56153#issuecomment-1773331706. I struggled with "When an enum initializer expression that isn't a string literal has a string type". What counts as "string literal" here? Just `""` or also parenthesized expressions or ...? I ended up copying the logic from the `evaluate` function (see below starting in line 45127).

The result of `evaluate` is currently used to determine if reverse mappings need to be created ([here via `getConstantValue`](https://github.com/microsoft/TypeScript/blob/01a51d2d017dd50d487835282b7e4a2a4f7300ab/src/compiler/transformers/ts.ts#L1876)). Without changing that, `evaluatesToStringLiteral` can only return `true` when `evaluate` returns a string.

I'd actually be in favor of changing that. See my comment https://github.com/microsoft/TypeScript/issues/56153#issuecomment-1849782437. I'd prefer if an enum did not generate reverse mappings for template expressions. Then we could just `return true` here.

I'm also open to just `return false` here in this PR.

Reply:
I think one of us isn’t understanding something specific about this `TemplateExpression` case. A “string literal” means literally something in single or double quotes, but you instead implemented “trivially syntactically determinable to evaluate to a string” which I think is fine and good. But by that definition, I don’t see why you don’t return `true` for cases like `` `foo${1}` `` or indeed even `` `foo${a}${b}${await import("c").blah}` ``. You’re guaranteed to get a string no matter what values get interpolated into the template spans. I expected just a `return true` here. Unless I’m missing something big!

Reply:
> I think one of us isn’t understanding something specific about this `TemplateExpression` case.

Yeah. I wouldn't be surprised if that was me. Let me start with my goal:

I want to ensure that the regular JS emit and ts.transpileModule emit are the same (or equivalent) for enums.

TS emits reverse mappings for template expression initializers if an expression inside can't be resolved to a constant string. Example [workbench](https://www.typescriptlang.org/dev/bug-workbench/?#code/PTAEAEDMEsBsFMB2BDAtvAXKAdMSB7fbAFwGcAoaVAB3wCdjQBvUAIQEEAlUAX1Ejr5UoAOS4ARsjoiA3OSQBXYQDFCzUO1ABeUAAMAJEw6ceu3uXIgIMBCnRYJUkhXgAPWg1ABjfIlKNjbVFJaRkgA):

```ts
import {BAR} from './bar';
enum Foo { A = `${BAR}` }
```

This emits `Foo["A"] = "bar";` if the `./bar` module is known. But with ts.transpileModule it emits ```Foo[Foo["A"] = `${BAR}`] = "A";```. This difference breaks a few tests in our code base.

If my understanding is correct so far, I think we have 2 options for template expressions here:

- `return false`: The more common (at least in our code base) NoSubstritionTemplateLiteral would still work. But TemplateExpressions would not be allowed under isolatedModules as enum member initializers.
- Some version of what I have in this PR.
- `return true` _and_ change TS emit so a TemplateExpression does not generate reverse mappings.

I'm open to all of those.

Reply:
Ah, nope, it was me who was confused. This makes sense. You need to ensure that the actual value of the string is statically knowable for a reverse mapping, not just that it’s going to be a string.

Reply:
Coming back to this. Maybe I’m still confused.
> ```ts
> import {BAR} from './bar';
> enum Foo { A = `${BAR}` }
> ```
> This emits `Foo["A"] = "bar";` if the `./bar` module is known. But with `ts.transpileModule` it emits `Foo[Foo["A"] = `${BAR}`] = "A";`.

I think the `transpileModule` output is bugged. Only numeric enum values are supposed to get reverse mappings, and you can syntactically tell that this is not a numeric value.

Although we typically don’t care much what emits from `tsc` when errors are present, fixing the `transpileModule` output here would probably involve “fixing” the full emit in the case where `BAR` doesn’t have a string literal type:

```ts
const BAR = 2..toFixed(0);

enum Foo {
    A = `${BAR}`
}
```

This is already an error today, because “computed” enum members must be numeric, and this has a string type. But the emit includes a reverse mapping:

```js
Foo[Foo["A"] = `${BAR}`] = "A";
```

I suspect this reverse mapping is produced for the same reason it’s produced in the check-error-free code transpiled with `ts.transpileModule`. The logic must be “I can’t evaluate this out to a known string value, so it’s considered computed, so it must be numeric, so it gets a reverse mapping. Any inconsistencies in this logic will be covered by a check error.” But that’s not true when the only reason a known string value couldn’t be determined was the fact that we’re transpiling.

If we change this logic to never produce a reverse mapping when the value can’t be fully statically determined but _can_ be statically determined to be a string, it feels like we can loosen the restrictions here.

Does that make sense?

Reply:
Yes, I think I can follow. It'd be nice to allow the example above, because with full type info we can determine a known string value.

> If we change this logic to never produce a reverse mapping when the value can’t be fully statically determined but can be statically determined to be a string, it feels like we can loosen the restrictions here.

Do we really loosen the restrictions if we do that? I can't think of one right now.

I added a commit to the end of the PR that removes the reverse mapping in case the enum initializer is syntactically a string. It didn't change the emit of any tests. That might be expected though because most tests have full type information?

How would you like to proceed? Would you like to discuss this internally?

Reply:
> Do we really loosen the restrictions if we do that? I can't think of one right now.

Yes, as seen when the error in the enumWithNonLiteralStringInitializer test went away, because now it’s possible for `ts.transpileModule` and `tsc` to produce behaviorally equivalent emit:

```ts
// tsc:
var A;
(function (A) {
    A["a"] = "2";
})(A || (A = {}));

// transpileModule:
var A;
(function (A) {
    A["a"] = `${helpers_1.foo}`;
})(A || (A = {}));
```

We would just have to ensure that the `transpileModule` emit is updated to this.

> How would you like to proceed? Would you like to discuss this internally?

Yeah, I’ll bring this to a design meeting soon.